src/mono/mono/metadata/metadata.c

```diff
@@ -5107,8 +5107,12 @@ mono_metadata_custom_attrs_from_index (MonoImage *meta, guint32 index)
 		return 0;
 
 	if (G_UNLIKELY (meta->has_updates)) {
-		if (!found && !mono_metadata_update_metadata_linear_search (meta, tdef, &loc, table_locator))
+		if (mono_metadata_table_num_rows (meta, MONO_TABLE_CUSTOMATTRIBUTE) > table_info_get_rows (tdef) || mono_metadata_update_has_modified_rows (tdef)) {
+			if (!found && !mono_metadata_update_metadata_linear_search (meta, tdef, &loc, table_locator))
+				return 0;
```

Comment:
I don't think we need `|| mono_metadata_update_has_modified_rows (tdef)` if the rows were modified, the binary search would find them - the linear search only really kicks in when there's added rows.

Also, we should make the whole thing conditional on not having found the custom attributes before.

So like this:
```suggestion
		if (!found && (mono_metadata_table_num_rows (meta, MONO_TABLE_CUSTOMATTRIBUTE) > table_info_get_rows (tdef))) {
			if (!mono_metadata_update_metadata_linear_search (meta, tdef, &loc, table_locator))
				return 0;
```